lib/silence_generator.ex

```diff
@@ -49,33 +49,61 @@ defmodule Membrane.SilenceGenerator do
   @impl true
   def handle_demand(:output, size, :bytes, _ctx, %{caps: caps} = state) do
     time = RawAudio.bytes_to_time(size, caps)
-    do_handle_demand(time, state)
+    do_handle_demand(1, time, state)
   end
 
   def handle_demand(:output, buffers, :buffers, _ctx, state) do
     %{caps: caps, frames_per_buffer: frames_per_buffer} = state
 
-    time = buffers * RawAudio.frames_to_time(frames_per_buffer, caps)
-    do_handle_demand(time, state)
+    time = RawAudio.frames_to_time(frames_per_buffer, caps)
+    do_handle_demand(buffers, time, state)
   end
 
+  defp do_handle_demand(buffers_number, time, state, buffers \\ [])
+
   defp do_handle_demand(
+         buffers_number,
          time,
-         %{caps: caps, duration: :infinity, passed_time: passed_time} = state
+         %{caps: caps, duration: :infinity, passed_time: passed_time} = state,
+         buffers
        ) do
     buffer = %Buffer{payload: RawAudio.silence(caps, time), pts: passed_time}
-    {{:ok, buffer: {:output, buffer}}, %{state | passed_time: passed_time + time}}
+    buffers = [buffer | buffers]
+    state = %{state | passed_time: passed_time + time}
+
+    if buffers_number == 1 do
+      {{:ok, buffer: {:output, Enum.reverse(buffers)}}, state}
+    else
+      do_handle_demand(buffers_number - 1, time, state, buffers)
+    end
   end
 
-  defp do_handle_demand(time, state) do
-    %{caps: caps, duration: duration, passed_time: passed_time} = state
+  defp do_handle_demand(
+         buffers_number,
+         time,
+         %{caps: caps, duration: duration, passed_time: passed_time} = state,
+         buffers
+       ) do
+    {buffer, state} =
+      if passed_time + time < duration do
+        {%Buffer{payload: RawAudio.silence(caps, time), pts: passed_time},
+         %{state | passed_time: passed_time + time}}
+      else
+        {%Buffer{payload: RawAudio.silence(caps, duration - passed_time), pts: passed_time},
+         %{state | passed_time: duration}}
+      end
 
-    if passed_time + time < duration do
-      buffer = %Buffer{payload: RawAudio.silence(caps, time), pts: passed_time}
-      {{:ok, buffer: {:output, buffer}}, %{state | passed_time: passed_time + time}}
-    else
-      buffer = %Buffer{payload: RawAudio.silence(caps, duration - passed_time), pts: passed_time}
-      {{:ok, buffer: {:output, buffer}, end_of_stream: :output}, %{state | passed_time: duration}}
+    buffers = [buffer | buffers]
+
+    case {state.passed_time, buffers_number} do
+      {^duration, _} ->
+        {{:ok, buffer: {:output, Enum.reverse(buffers)}, end_of_stream: :output}, state}
+
+      {_, 1} ->
+        {{:ok, buffer: {:output, Enum.reverse(buffers)}}, state}
+
+      _else ->
+        do_handle_demand(buffers_number - 1, time, state, buffers)
```

Comment:
There's some code duplication here and the recursion mixed with other logic makes it all complicated. The only thing to do is to generate many smaller buffers instead a single big one, so we could just have something like `defp generate_buffers(start_time, chunk_time, total_time)` and let it generate buffers instead of returning a single buffer.